server/src/main/java/org/opensearch/action/support/TimeoutTaskCancellationUtility.java

```diff
@@ -0,0 +1,163 @@
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.cluster.node.tasks.cancel.CancelTasksRequest;
+import org.opensearch.client.OriginSettingClient;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.search.SearchService;
+import org.opensearch.tasks.CancellableTask;
+import org.opensearch.tasks.TaskId;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.opensearch.action.admin.cluster.node.tasks.get.GetTaskAction.TASKS_ORIGIN;
+import static org.opensearch.action.search.TransportSearchAction.SEARCH_CANCEL_AFTER_TIME_INTERVAL_SETTING;
+
+public class TimeoutTaskCancellationUtility {
+
+    private static final Logger logger = LogManager.getLogger(TimeoutTaskCancellationUtility.class);
+
+    /**
+     * Wraps a listener with a timeout listener {@link TimeoutRunnableListener} to schedule the task cancellation for provided tasks on
+     * generic thread pool
+     * @param client - {@link NodeClient}
+     * @param taskToCancel - task to schedule cancellation for
+     * @param clusterSettings - {@link ClusterSettings}
+     * @param listener - original listener associated with the task
+     * @return wrapped listener
+     */
+    public static <Response> ActionListener<Response> wrapWithCancellationListener(
+        NodeClient client,
+        CancellableTask taskToCancel,
+        ClusterSettings clusterSettings,
+        ActionListener<Response> listener
+    ) {
+        final TimeValue globalTimeout = clusterSettings.get(SEARCH_CANCEL_AFTER_TIME_INTERVAL_SETTING);
+        final TimeValue timeoutInterval = (taskToCancel.getCancellationTimeout() == null)
+            ? globalTimeout
+            : taskToCancel.getCancellationTimeout();
+        // Note: -1 (or no timeout) will help to turn off cancellation. The combinations will be request level set at -1 or request level
+        // set to null and cluster level set to -1.
+        ActionListener<Response> listenerToReturn = listener;
+        if (timeoutInterval.equals(SearchService.NO_TIMEOUT)) {
+            return listenerToReturn;
+        }
+
+        try {
+            final TimeoutRunnableListener<Response> wrappedListener = new TimeoutRunnableListener<>(timeoutInterval, listener, () -> {
+                final CancelTasksRequest cancelTasksRequest = new CancelTasksRequest();
+                cancelTasksRequest.setTaskId(new TaskId(client.getLocalNodeId(), taskToCancel.getId()));
+                cancelTasksRequest.setReason("Cancellation timeout of " + timeoutInterval + " is expired");
+                // force the origin to execute the cancellation as a system user
+                new OriginSettingClient(client, TASKS_ORIGIN).admin()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *OriginSettingClient* constructor in a try-finally block. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).